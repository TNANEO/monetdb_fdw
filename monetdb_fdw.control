# monetdb_fdw extension
comment = 'a monetdb foreign-data wrapper'
default_version = '0.0'
module_pathname = '$libdir/monetdb_fdw'
relocatable = true

