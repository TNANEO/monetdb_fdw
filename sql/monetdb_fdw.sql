CREATE FUNCTION monetdb_fdw_handler()
RETURNS fdw_handler
AS 'monetdb_fdw'
LANGUAGE C STRICT;

CREATE FUNCTION monetdb_fdw_validator(text[], oid)
RETURNS void
AS 'monetdb_fdw'
LANGUAGE C STRICT;

CREATE FOREIGN DATA WRAPPER monetdb_fdw
  HANDLER monetdb_fdw_handler
  VALIDATOR monetdb_fdw_validator;

CREATE SERVER monetdb_server FOREIGN DATA WRAPPER monetdb_fdw;
CREATE USER MAPPING FOR current_user SERVER monetdb_server;

CREATE FOREIGN TABLE nation (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'dbt3', table 'nation')
;

SELECT * FROM nation ORDER BY 1;

CREATE FOREIGN TABLE nation1 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'nosuchhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'dbt3', table 'nation')
;

SELECT * FROM nation1;

CREATE FOREIGN TABLE nation2 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50001', user 'monetdb', passwd 'monetdb', dbname 'dbt3', table 'nation')
;

SELECT * FROM nation2;

CREATE FOREIGN TABLE nation3 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'nosuchuser', passwd 'monetdb', dbname 'dbt3', table 'nation')
;

SELECT * FROM nation3;

CREATE FOREIGN TABLE nation4 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'wrongpass', dbname 'dbt3', table 'nation')
;

SELECT * FROM nation4;

CREATE FOREIGN TABLE nation5 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'nosuchdb', table 'nation')
;

SELECT * FROM nation5;

CREATE FOREIGN TABLE nation6 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'dbt3', table 'nosuchtable')
;

SELECT * FROM nation6;

CREATE FOREIGN TABLE nation7 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'dbt3', table 'select * fro nation')
;

SELECT * FROM nation7;

CREATE FOREIGN TABLE nation8 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'dbt3', query 'select * fro nation')
;

SELECT * FROM nation8;

DROP FOREIGN TABLE nation1;
DROP FOREIGN TABLE nation2;
DROP FOREIGN TABLE nation3;
DROP FOREIGN TABLE nation4;
DROP FOREIGN TABLE nation5;
DROP FOREIGN TABLE nation6;
DROP FOREIGN TABLE nation7;
DROP FOREIGN TABLE nation8;

\d
