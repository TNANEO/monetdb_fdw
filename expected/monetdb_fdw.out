CREATE FUNCTION monetdb_fdw_handler()
RETURNS fdw_handler
AS 'monetdb_fdw'
LANGUAGE C STRICT;
CREATE FUNCTION monetdb_fdw_validator(text[], oid)
RETURNS void
AS 'monetdb_fdw'
LANGUAGE C STRICT;
CREATE FOREIGN DATA WRAPPER monetdb_fdw
  HANDLER monetdb_fdw_handler
  VALIDATOR monetdb_fdw_validator;
CREATE SERVER monetdb_server FOREIGN DATA WRAPPER monetdb_fdw;
CREATE USER MAPPING FOR current_user SERVER monetdb_server;
CREATE FOREIGN TABLE nation (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'dbt3', table 'nation')
;
SELECT * FROM nation ORDER BY 1;
 n_nationkey |          n_name           | n_regionkey |                                                    n_comment                                                     
-------------+---------------------------+-------------+------------------------------------------------------------------------------------------------------------------
           0 | ALGERIA                   |           0 | final accounts wake quickly. special reques
           1 | ARGENTINA                 |           1 | idly final instructions cajole stealthily. regular instructions wake carefully blithely express accounts. fluffi
           2 | BRAZIL                    |           1 | always pending pinto beans sleep sil
           3 | CANADA                    |           1 | foxes among the bold requests
           4 | EGYPT                     |           4 | pending accounts haggle furiously. furiously bold accounts detect. platelets at the packages haggle caref
           5 | ETHIOPIA                  |           0 | fluffily ruthless requests integrate fluffily. pending ideas wake blithely acco
           6 | FRANCE                    |           3 | even requests detect near the pendin
           7 | GERMANY                   |           3 | blithely ironic foxes grow. quickly pending accounts are b
           8 | INDIA                     |           2 | ironic packages should have to are slyly around the special, ironic accounts. iron
           9 | INDONESIA                 |           2 | unusual excuses are quickly requests. slyly ironic accounts haggle carefully above the pendin
          10 | IRAN                      |           4 | blithely even accounts about the furiously regular foxes nag slyly final accounts. quickly final fo
          11 | IRAQ                      |           4 | express, pending deposits boost quick
          12 | JAPAN                     |           2 | blithely final packages cajole quickly even dependencies? blithely regular deposits haggle express, ironic re
          13 | JORDAN                    |           4 | blithe, express deposits boost carefully busy accounts. furiously pending depos
          14 | KENYA                     |           0 | ironic requests boost. quickly pending pinto beans cajole slyly slyly even deposits. ironic packages 
          15 | MOROCCO                   |           0 | ideas according to the fluffily final pinto beans sleep furiously
          16 | MOZAMBIQUE                |           0 | ironic courts wake fluffily even, bold deposi
          17 | PERU                      |           1 | final, final accounts sleep slyly across the requests. 
          18 | CHINA                     |           2 | bold accounts are. slyly ironic escapades haggle acc
          19 | ROMANIA                   |           3 | deposits boost against the brave id
          20 | SAUDI ARABIA              |           4 | fluffily final accounts wake slyly-- fi
          21 | VIETNAM                   |           2 | doggedly ironic requests haggle furiously ironic, ironic packages. furiously final courts wake fur
          22 | RUSSIA                    |           3 | slowly pending patterns x-ray quickly. ironic, even accounts haggle furiously. even, final deposits mold bl
          23 | UNITED KINGDOM            |           3 | fluffily regular pinto beans breach according to the ironic dolph
          24 | UNITED STATES             |           1 | blithely regular deposits serve furiously blithely regular warthogs! slyly fi
(25 rows)

CREATE FOREIGN TABLE nation1 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'nosuchhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'dbt3', table 'nation')
;
SELECT * FROM nation1;
ERROR:  monetdb_fdw: getaddrinfo failed: Name or service not known
CREATE FOREIGN TABLE nation2 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50001', user 'monetdb', passwd 'monetdb', dbname 'dbt3', table 'nation')
;
SELECT * FROM nation2;
ERROR:  monetdb_fdw: could not connect to localhost:50001: Connection refused
CREATE FOREIGN TABLE nation3 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'nosuchuser', passwd 'monetdb', dbname 'dbt3', table 'nation')
;
SELECT * FROM nation3;
ERROR:  monetdb_fdw: InvalidCredentialsException:checkCredentials:invalid credentials for user 'nosuchuser'

CREATE FOREIGN TABLE nation4 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'wrongpass', dbname 'dbt3', table 'nation')
;
SELECT * FROM nation4;
ERROR:  monetdb_fdw: InvalidCredentialsException:checkCredentials:invalid credentials for user 'monetdb'

CREATE FOREIGN TABLE nation5 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'nosuchdb', table 'nation')
;
SELECT * FROM nation5;
ERROR:  monetdb_fdw: monetdbd: no such database 'nosuchdb', please create it first

CREATE FOREIGN TABLE nation6 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'dbt3', table 'nosuchtable')
;
SELECT * FROM nation6;
ERROR:  monetdb_fdw: 42S02!SELECT: no such table 'nosuchtable'

CONTEXT:  relation nation6, line 0
CREATE FOREIGN TABLE nation7 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'dbt3', table 'select * fro nation')
;
SELECT * FROM nation7;
ERROR:  monetdb_fdw: 42000!syntax error, unexpected SELECT in: "select * from select"

CONTEXT:  relation nation7, line 0
CREATE FOREIGN TABLE nation8 (
        "n_nationkey" INTEGER,
        "n_name"      CHAR(25),
        "n_regionkey" INTEGER,
        "n_comment"   VARCHAR(152)
) SERVER monetdb_server
OPTIONS (host 'localhost', port '50000', user 'monetdb', passwd 'monetdb', dbname 'dbt3', query 'select * fro nation')
;
SELECT * FROM nation8;
ERROR:  monetdb_fdw: 42000!syntax error, unexpected IDENT, expecting SCOLON in: "select * fro"

CONTEXT:  relation nation8, line 0
DROP FOREIGN TABLE nation1;
DROP FOREIGN TABLE nation2;
DROP FOREIGN TABLE nation3;
DROP FOREIGN TABLE nation4;
DROP FOREIGN TABLE nation5;
DROP FOREIGN TABLE nation6;
DROP FOREIGN TABLE nation7;
DROP FOREIGN TABLE nation8;
\d
            List of relations
 Schema |  Name  |     Type      | Owner 
--------+--------+---------------+-------
 public | nation | foreign table | snaga
(1 row)

